api/sendfild/service/waitmanager/waitmanager.go

```diff
@@ -0,0 +1,152 @@
+package waitmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/textileio/powergate/v2/lotus"
+	"github.com/textileio/textile/v2/api/sendfild/service/store"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+var log = logging.Logger("sendfil-waitmanager")
+
+type WaitManager struct {
+	clientBuilder lotus.ClientBuilder
+	store         *store.Store
+	confidence    uint64
+	waitTimeout   time.Duration
+	waiting       map[primitive.ObjectID]*WaitRunner
+	waitingLck    sync.Mutex
+	ticker        *time.Ticker
+	mainCtx       context.Context
+	mainCtxCancel context.CancelFunc
+}
+
+func New(cb lotus.ClientBuilder, store *store.Store, confidence uint64, waitTimeout time.Duration, retryWaitFrequency time.Duration) (*WaitManager, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	w := &WaitManager{
+		clientBuilder: cb,
+		store:         store,
+		confidence:    confidence,
+		waitTimeout:   waitTimeout,
+		waiting:       make(map[primitive.ObjectID]*WaitRunner),
+		ticker:        time.NewTicker(retryWaitFrequency),
+		mainCtx:       ctx,
+		mainCtxCancel: cancel,
+	}
+
+	if err := w.waitAllPending(ctx, true); err != nil {
+		cancel()
+		return nil, fmt.Errorf("calling waitAllPending: %v", err)
+	}
+
+	w.bindTicker(ctx)
+
+	return w, nil
+}
+
+func (w *WaitManager) RegisterTxn(objID primitive.ObjectID, messageCid string) error {
+	runner, err := w.getOrCreateRunner(objID, messageCid)
+	if err != nil {
+		return fmt.Errorf("getting wait runner: %v", err)
+	}
+	runner.Start()
+	return nil
+}
+
+func (w *WaitManager) Subscribe(objID primitive.ObjectID, messageCid string, listener chan WaitResult) (CancelListenerFunc, error) {
+	runner, err := w.getOrCreateRunner(objID, messageCid)
+	if err != nil {
+		return nil, fmt.Errorf("getting wait runner: %v", err)
+	}
+	cancel := runner.AddListener(listener)
+	runner.Start()
+	return cancel, nil
+}
```

Comment:
These two funcs are similar but `Subscribe` adds the provided listener before starting the runner and of course returns a cancelation function.